torch/utils/_pytree.py

```diff
@@ -26,194 +28,128 @@
 To improve the performance we can move parts of the implementation to C++.
 """
 
+DEFAULT_TREESPEC_SERIALIZATION_PROTOCOL = 1
+
+Context = Any
+PyTree = Any
+FlattenFunc = Callable[[PyTree], Tuple[List, Context]]
+UnflattenFunc = Callable[[List, Context], PyTree]
+DumpableContext = Any  # Any json dumpable text
+ToDumpableContextFn = Callable[[Context], DumpableContext]
+FromDumpableContextFn = Callable[[DumpableContext], Context]
+
 # A NodeDef holds two callables:
 # - flatten_fn should take the collection and return a flat list of values.
 #   It can also return some context that is used in reconstructing the
 #   collection.
 # - unflatten_fn should take a flat list of values and some context
 #   (returned by flatten_fn). It returns the collection by reconstructing
 #   it from the list and the context.
-# - to_str_fn takes a TreeSpec with the specific type and a list of its children
-#   TreeSpecs already converted to strings, and returns a string representation
-#   of this TreeSpec
-# - maybe_from_str_fn takes in a string and if this string represents a TreeSpec
-#   of this type, returns the type, the context, and a string representation of
-#   its children specs. Otherwise it returns None.
-Context = Any
-PyTree = Any
-FlattenFunc = Callable[[PyTree], Tuple[List, Context]]
-UnflattenFunc = Callable[[List, Context], PyTree]
-ToStrFunc = Callable[["TreeSpec", List[str]], str]
-MaybeFromStrFunc = Callable[[str], Optional[Tuple[Any, Context, str]]]
-
 class NodeDef(NamedTuple):
     type: Type[Any]
     flatten_fn: FlattenFunc
     unflatten_fn: UnflattenFunc
-    to_str_fn: ToStrFunc
-    maybe_from_str_fn: MaybeFromStrFunc
 
 SUPPORTED_NODES: Dict[Type[Any], NodeDef] = {}
 
+# _SerializeNodeDef holds two callables:
+# - to_dumpable_context takes a TreeSpec, and returns a serialized string format of the
+#   context, and the version number
+# - from_dumpable_context takes in a string representation of the context, and the
+#   version, and returns the deserialized context
+class _SerializeNodeDef(NamedTuple):
+    typ: Type[Any]
+    serialized_type: str
+    to_dumpable_context: Optional[ToDumpableContextFn]
+    from_dumpable_context: Optional[FromDumpableContextFn]
+
+SUPPORTED_SERIALIZED_TYPES: Dict[Type[Any], _SerializeNodeDef] = {}
+SERIALIZED_TYPE_TO_PYTHON_TYPE: Dict[str, Type[Any]] = {}
+
 def _register_pytree_node(
     typ: Any,
     flatten_fn: FlattenFunc,
     unflatten_fn: UnflattenFunc,
-    to_str_fn: Optional[ToStrFunc] = None,
-    maybe_from_str_fn: Optional[MaybeFromStrFunc] = None,
+    *,
+    to_dumpable_context: Optional[ToDumpableContextFn] = None,
+    from_dumpable_context: Optional[FromDumpableContextFn] = None,
```

Comment:
Can you add a docstring while you're at it to `_register_pytree_node`? Or a quick comment here to say that `to_dumpable_context/from_dumpable_context` are optional. When people call `_register_pytree_node` they actually end up providing everything; we should let them know that they "only need to provide these two if they want custom json serialization, which is only being used in torch.export right now".